packages/cozy-pouch-link/src/loop.js

```diff
@@ -0,0 +1,139 @@
+const sleep = delay => new Promise(resolve => setTimeout(resolve, delay))
+
+/**
+ * Utility to call a function (task) periodically
+ * and on demand immediately.
+ *
+ * Public API
+ *
+ * - start
+ * - stop
+ * - scheduleImmediateTask
+ * - waitForCurrentTask
+ */
+class Loop {
+  constructor(task, delay, _afterRound, _sleep) {
+    this.task = task
+    this.delay = delay
+
+    this.round = this.round.bind(this)
+    this.immediateTasks = []
+    this.started = false
+
+    // Only for tests
+    this.afterRound = _afterRound
+    this.sleep = _sleep || sleep
+  }
+
+  /**
+   * Starts the loop. Will run the task periodically each `this.delay` ms.
+   * Ignores multiple starts.
+   */
+  start() {
+    if (this.started) {
+      return
+    }
+    this.round()
+    this.started = true
+  }
+
+  /**
+   * Stops the loop, clears immediate tasks.
+   * Cancels current task if possible
+   */
+  stop() {
+    if (!this.started) {
+      return
+    }
+    if (this.currentTask && this.currentTask.cancel) {
+      this.currentTask.cancel()
+    }
+    clearTimeout(this.timeout)
+    this.clearImmediateTasks()
+    this.started = false
+  }
+
+  async waitForCurrent() {
+    await this.currentTask
+    return
+  }
+
+  /**
+   * Flushes the immediate tasks list and calls each task.
+   * Each task is awaited before the next is started.
+   */
+  async runImmediateTasks() {
+    while (this.immediateTasks.length > 0) {
+      const task = this.immediateTasks.splice(0, 1)[0]
+      await this.runTask(task)
+    }
+  }
+
+  /**
+   * If loop started, schedules tasks that will be run in priority without
+   * delay at next round.
+   */
+  async scheduleImmediateTask(task) {
```

Comment:
If `task`is optional (it seems to be), maybe state it explicitly : `async scheduleImmediateTask(task=null)`